File: 2021/python2021/aoc/day01.py
#!/usr/bin/env python
"""
Advent Of Code 2021 Day 1
https://adventofcode.com/2021/day/1
"""
from typing import List


def parse(filename: str) -> List[int]:
    with open(filename) as file:
        return [int(line.strip()) for line in file.readlines()]


def increase_count(data: List[int], window: int) -> int:
    return sum(b > a for (a, b) in zip(data, data[window:]))


class Day01:
    """AoC 2021 Day 01"""

    @staticmethod
    def part1(filename: str) -> int:
        data = parse(filename)
        return increase_count(data, 1)

    @staticmethod
    def part2(filename: str) -> int:
        data = parse(filename)
        return increase_count(data, 3)

File: 2021/elixir2021/lib/day01.ex
defmodule Elixir2021.Day01 do
  def parse(filename) do
    File.stream!(filename)
    |> Stream.map(&String.trim/1)
    |> Stream.map(&String.to_integer/1)
  end

  def increase_count(nums, window) do
    nums
    |> Enum.zip(Enum.drop(nums, window))
    |> Enum.reduce(0, fn {a, b}, acc ->
      if b > a, do: acc + 1, else: acc
    end)
  end

  def part1(filename) do
    parse(filename)
    |> increase_count(1)
  end

  def part2(filename) do
    parse(filename)
    |> increase_count(3)
  end
end
